infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java

```diff
@@ -21,14 +32,123 @@
+     */
+    default void sendWrongClusterIdMessage(CorfuMsg msg, ChannelHandlerContext ctx, UUID currentClusterID) {
+        sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_CLUSTER_ID, currentClusterID));
+        log.warn("Incoming message with a wrong cluster ID, got {}, expected {}, message was {}",
+                msg.getClusterID(), currentClusterID, msg);
+    }
+
+    /**
+     * Validate the epoch of a CorfuMsg, and send a WRONG_EPOCH response if
+     * the server is in the wrong epoch. Ignored if the message type is reset (which
+     * is valid in any epoch).
+     *
+     * @param msg The incoming message to validate.
+     * @param ctx The context of the channel handler.
+     * @return True, if the epoch is correct, but false otherwise.
+     */
+    default boolean epochIsValid(CorfuMsg msg, ChannelHandlerContext ctx) {
+        long serverEpoch = getServerEpoch();
+        if (msg.getEpoch() != serverEpoch) {
+            sendWrongEpochMessage(msg, ctx);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Validate that the message's cluster ID is equal to the cluster ID of a bootstrapped layout.
+     *
+     * @param msg           The incoming message.
+     * @param ctx           The context of the channel handler.
+     * @param currentLayout The layout a server was bootstrapped with.
+     * @return True, if the message's cluster ID is equal to this node's cluster ID, otherwise false.
+     */
+    default boolean clusterIdIsValid(CorfuMsg msg, ChannelHandlerContext ctx, Layout currentLayout) {
+        UUID currentClusterID = currentLayout.getClusterId();
+
+        boolean clusterIdsMatch = msg.getClusterID()
+                .equals(currentClusterID);
+
+        if (!clusterIdsMatch) {
+            sendWrongClusterIdMessage(msg, ctx, currentClusterID);
+        }
+        return clusterIdsMatch;
+    }
+
+    /**
+     * Validate the incoming message. The message is valid if:
+     * 1) The flag ignoreEpoch is set to true.
+     * 2) Otherwise,
+     *      a. The message's epoch is the same as the router's epoch.
+     *      b. The current layout server should be bootstrapped and
+     *         the message's cluster ID should be equal to the bootstrapped layout's cluster ID.
+     *
+     * @param msg The incoming message.
+     * @param ctx The context of the channel handler.
+     * @return True, if it's a valid message, and false otherwise.
+     */
+    default boolean messageIsValid(CorfuMsg msg, ChannelHandlerContext ctx) {
+        if (msg.getMsgType().ignoreEpoch) {
+            return true;
+        }
+
+        if (!epochIsValid(msg, ctx)) {
+            return false;
+        }
+
+        return getCurrentLayout()
+                .map(layout -> clusterIdIsValid(msg, ctx, layout))
+                .orElseGet(() -> {
+                    sendNoBootstrapMessage(msg, ctx);
+                    return false;
+                });
```

Comment:
This seems incorrect. 

Imagine a fault detector from one cluster trying to seal a node from another cluster. Since a seal message is marked as ignore epoch, the second check won't be executed and thus allowing fault detectors from different cluster to reconfigure each other's nodes.

Reply:
Resolving for now, as this situation is impossible since a fault detector's management runtime will not start trying to connect to another cluster, failing the clusterId check.

Reply:
Which clusterId check are referring to ? `messageIsValid` isn't sufficient for RPCs that ignore epoch checks, because ignore epoch is within a context of a cluster, so returning true only based on that flag is incorrect.

I'm thinking of the following scenario: The management agent has a layout and connected to 3 nodes (A, B and C). Since reset is an ad-hoc call it can happen at any time. Now, some one resets node C, the netty channel will auto reconnect by-passing all checks. Now the failure detector sees that C node doesn't have the latest layout and tries to patch it with LAYOUT_COMMITTED (which will pass messageIsValid because it sets ignoreEpoch=true). 

This is incorrect because C got re-added to the cluster after being reset. First of all why cant this scenario happen ? and at what point would message validation fail for the next step (i.e. LAYOUT_COMMITTED). 

Obviously, this isn't an issue for msgs that don't ignore epochs and its not specific to LAYOUT_COMMITTED, but all messages that choose to ignore epoch checks.

Reply:
I see, I did not know that the netty channel reconnects bypassing the clusterId checks. This way, FD can interfere with the workflows that have resets. In that case, the correct behavior would be to ignore the epochIsValid check for the messages that have the ignoreEpoch=True but continue checking the clusterId nonetheless.

Reply:
@Maithem 
I fixed the issue related to checking the layout and clusterId for the ignoreEpoch messages, however it introduced another issue that these checks should actually be ignored for some of the message types. This makes sense because these message types have to deal with pings, cluster discovery, cluster bootstrapping or ADD_NODE workflow related (A cluster should not/might not be bootstrapped, when the server router receives these messages). I also had to fix some of the tests accordingly. Let me know if you got any questions, and please review when you get the chance.